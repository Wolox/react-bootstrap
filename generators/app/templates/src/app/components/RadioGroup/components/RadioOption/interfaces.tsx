import React from 'react';

export interface IRadioOption {
  checked?: boolean;
  children: React.FunctionComponent;
  className?: string;
  disabled?: boolean;
  key: string | number;
  inputClassName?: string;
  labelClassName?: string;
  label?: string | undefined;
  name?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  value: string;
  [rest: string]: any; 
}
