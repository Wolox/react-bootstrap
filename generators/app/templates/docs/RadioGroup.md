## RadioGroup

 **Props**

| Property | Type | Required? | Default | Description |
|:---|:---:|:---:|:---:|:---|
| disabled | Boolean | | `false` | Controls if all the options are disabled |
| className | String | | `''` | Styles to apply to the radio inputs container |
| name | String | ✓ | | If associated to a form, the name is submitted with the control's value as part of the form data. It is passed to every `RadioOption` child |
| children | ReactNode[] | ✓ | | Array of options in the group, which typically would be RadioOption, but it could be any component. The data received from the RadioGroup comes from RadioGroupContext |
| onChange | (event: ChangeEvent<HTMLInputElement>) => void | ✓ | | Function to call on any option selection change |
| initialValue | String | | `''` | Specifies the initial selected value on mounting |

**RadioOption props**

| Property | Type | Required? | Default | Description |
|:---|:---:|:---:|:---:|:---|
| children | ReactNode | | | children which will be rendered after the input and label inside the option |
| className | String | | `''` | Styles to apply to the input and label container |
| inputClassName | String | | `''` | Styles to apply to the input |
| labelClassName | String | | `''` | Styles to apply to the label |
| label | String | | `''` | Input associated label |
| id | String | ✓ | | The value the option will have if selected |
| disabled | Boolean | | `false` | Controls if the option is available to select |

**Usage**

 ```jsx
  <RadioGroup onChange={onChange} initialValue={selectedOption} name="radio-button">
    <RadioOption id="option1" label="Option 1" />
    <RadioOption id="option2" label="Option 2" />
  </RadioGroup>
```
