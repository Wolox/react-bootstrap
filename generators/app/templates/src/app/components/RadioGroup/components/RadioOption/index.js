import React from 'react';
import PropTypes from 'prop-types';

function RadioOption({ className, inputClassName, labelClassName, label, children, ...rest }) {
  return (
    <div className={className}>
      {/* eslint-disable-next-line react/forbid-dom-props */}
      <input type="radio" id={label} className={inputClassName} {...rest} />
      {label && (
        <label htmlFor={label} className={labelClassName}>
          {label}
        </label>
      )}
      {children}
    </div>
  );
}

RadioOption.propTypes = {
  value: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  inputClassName: PropTypes.string,
  label: PropTypes.string,
  labelClassName: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func
};

export default RadioOption;
