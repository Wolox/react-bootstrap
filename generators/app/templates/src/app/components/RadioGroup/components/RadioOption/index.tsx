import React, { useContext } from 'react';

import RadioGroupContext from '~components/RadioGroup/context';

export interface RadioOptionProps {
  children?: React.ReactNode;
  className?: string;
  disabled?: boolean;
  inputClassName?: string;
  labelClassName?: string;
  label: string;
  id: string;
}

function RadioOption({
  children,
  className = '',
  inputClassName = '',
  labelClassName = '',
  label,
  id,
  disabled = false,
  ...props
}: RadioOptionProps) {
  const { name, handleChange, groupDisabled, selectedValue } = useContext(RadioGroupContext);
  return (
    <div className={className}>
      <input
        type="radio"
        id={id}
        name={name}
        className={inputClassName}
        disabled={disabled || groupDisabled}
        value={id}
        onChange={handleChange}
        checked={selectedValue === id}
        data-testid={`${name}-${id}`}
        {...props}
      />
      <label htmlFor={name} className={labelClassName}>
        {label}
      </label>
      {children}
    </div>
  );
}

export default RadioOption;
