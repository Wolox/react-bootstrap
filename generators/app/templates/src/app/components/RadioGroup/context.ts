import { createContext } from 'react';

interface ContextValues {
  name: string;
  handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  groupDisabled: boolean;
  selectedValue: string;
}

const RadioGroupContext = createContext<ContextValues>({
  name: '',
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  handleChange: () => {},
  groupDisabled: false,
  selectedValue: ''
});

export default RadioGroupContext;
