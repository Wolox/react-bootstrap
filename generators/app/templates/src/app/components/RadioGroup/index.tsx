import React, { useState } from 'react';

import RadioGroupContext from '~components/RadioGroup/context';

export interface Props {
  disabled?: boolean;
  className?: string;
  name: string;
  children: React.ReactNode[];
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  initialValue: string;
}

function RadioGroup({
  disabled = false,
  className = '',
  name,
  children,
  onChange,
  initialValue = ''
}: Props) {
  const [selectedValue, setSelectedValue] = useState(initialValue);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSelectedValue(event.target.value);

    if (onChange) {
      onChange(event);
    }
  };

  return (
    <div className={className}>
      <RadioGroupContext.Provider value={{ selectedValue, handleChange, name, groupDisabled: disabled }}>
        {children}
      </RadioGroupContext.Provider>
    </div>
  );
}

export default RadioGroup;
