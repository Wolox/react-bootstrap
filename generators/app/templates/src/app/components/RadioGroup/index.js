import React, { Component, createContext } from 'react';
import PropTypes from 'prop-types';

import RadioOption from './components/RadioOption';

const { Provider, Consumer } = createContext();

class RadioGroup extends Component {
  static option = props => (
    <Consumer>
      {({ selectedValue, name, disabled, onChange }) => (
        <RadioOption
          {...props}
          name={name}
          onChange={onChange}
          checked={selectedValue === props.value}
          disabled={disabled || props.disabled}
        />
      )}
    </Consumer>
  );

  state = { selectedValue: this.props.selectedValue };

  handleChange = event => {
    const { onChange } = this.props;

    this.setState({ selectedValue: event.target.value });
    if (onChange) {
      onChange(event);
    }
  };

  render() {
    const { className, name, disabled, children } = this.props;
    const { selectedValue } = this.state;
    const context = { selectedValue, name, disabled, onChange: this.handleChange };

    return (
      <div className={className}>
        <Provider value={context}>{children}</Provider>
      </div>
    );
  }
}

RadioGroup.defaultProps = {
  className: '',
  disabled: false
};

RadioGroup.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  name: PropTypes.string,
  selectedValue: PropTypes.string,
  onChange: PropTypes.func
};

export default RadioGroup;
