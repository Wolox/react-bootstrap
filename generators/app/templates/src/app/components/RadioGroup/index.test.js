import React from 'react';
import { mount } from 'enzyme';

import RadioOption from './components/RadioOption';

import RadioGroup from './index';

describe('#RadioGroup', () => {
  let selectedOption = null;
  let component = null;
  const onChange = event => (selectedOption = event.target.value);
  const getInput = (comp, id) => comp.find(`[data-testid="radioButton-${id}"]`);

  beforeEach(() => {
    selectedOption = 'option2';
  });

  describe('when the group is disabled', () => {
    beforeEach(() => {
      component = mount(
        <RadioGroup onChange={onChange} initialValue={selectedOption} name="radioButton" disabled>
          <RadioOption id="option1" label="Option 1" />
          <RadioOption id="option2" label="Option 2" />
        </RadioGroup>
      );
    });

    it('all the inputs are disabled', () => {
      const selectedInput1 = getInput(component, 'option1');
      const selectedInput2 = getInput(component, 'option2');
      expect(selectedInput1.props().disabled).toBeTruthy();
      expect(selectedInput2.props().disabled).toBeTruthy();
    });

    it('matches last snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when an option is selected', () => {
    beforeEach(() => {
      component = mount(
        <RadioGroup onChange={onChange} initialValue={selectedOption} name="radioButton" disabled={false}>
          <RadioOption id="option1" label="Option 1" />
          <RadioOption id="option2" label="Option 2" />
        </RadioGroup>
      );
    });

    it('provides the selected value to the onChange callback', () => {
      const selectedInput = getInput(component, 'option1');
      selectedInput.simulate('change', { target: { value: 'option1' } });
      expect(selectedOption).toEqual('option1');
    });

    it('matches last snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
