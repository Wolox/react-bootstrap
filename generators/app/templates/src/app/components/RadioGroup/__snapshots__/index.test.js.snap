// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#RadioGroup when an option is selected matches last snapshot 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <input
      checked={false}
      className=""
      data-testid="radioButton-option1"
      disabled={false}
      id="option1"
      name="radioButton"
      onChange={[Function]}
      type="radio"
      value="option1"
    />
    <label
      className=""
      htmlFor="radioButton"
    >
      Option 1
    </label>
  </div>
  <div
    className=""
  >
    <input
      checked={true}
      className=""
      data-testid="radioButton-option2"
      disabled={false}
      id="option2"
      name="radioButton"
      onChange={[Function]}
      type="radio"
      value="option2"
    />
    <label
      className=""
      htmlFor="radioButton"
    >
      Option 2
    </label>
  </div>
</div>
`;

exports[`#RadioGroup when the group is disabled matches last snapshot 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <input
      checked={false}
      className=""
      data-testid="radioButton-option1"
      disabled={true}
      id="option1"
      name="radioButton"
      onChange={[Function]}
      type="radio"
      value="option1"
    />
    <label
      className=""
      htmlFor="radioButton"
    >
      Option 1
    </label>
  </div>
  <div
    className=""
  >
    <input
      checked={true}
      className=""
      data-testid="radioButton-option2"
      disabled={true}
      id="option2"
      name="radioButton"
      onChange={[Function]}
      type="radio"
      value="option2"
    />
    <label
      className=""
      htmlFor="radioButton"
    >
      Option 2
    </label>
  </div>
</div>
`;
